types/middlewares.d.ts

```diff
@@ -1,8 +1,16 @@
 declare module 'mongoose' {
 
-  type MongooseDocumentMiddleware = 'validate' | 'save' | 'remove' | 'updateOne' | 'deleteOne' | 'init';
-  type MongooseQueryMiddleware = 'count' | 'estimatedDocumentCount' | 'countDocuments' | 'deleteMany' | 'deleteOne' | 'distinct' | 'find' | 'findOne' | 'findOneAndDelete' | 'findOneAndRemove' | 'findOneAndReplace' | 'findOneAndUpdate' | 'remove' | 'replaceOne' | 'update' | 'updateOne' | 'updateMany';
-  type DocumentOrQueryMiddleware = 'updateOne' | 'deleteOne' | 'remove';
+  type MongooseQueryAndDocumentMiddleware = 'remove' | 'updateOne' | 'deleteOne';
```

Comment:
`validate` is another one that can be both Query and Document middleware

Reply:
According to https://mongoosejs.com/docs/middleware.html, validate is only a document middleware.

Reply:
We added a `Query.validate()` method. Mongoose calls this method when `runValidators` is set on a query, for example:

```javascript
'use strict';

const mongoose = require('mongoose');

const schema = new mongoose.Schema({ name: String });

schema.pre('validate', () => {
  console.log('Regular pre validate');
});
schema.pre('validate', { query: true, document: false }, () => {
  console.log('Pre query validate');
});

const Test = mongoose.model('Test', schema);

run().catch(err => console.log(err));

async function run() {
  await mongoose.connect('mongodb://localhost:27017/test');

  const doc = new Test();
  await doc.validate();

  console.log('----');
  await Test.updateOne({}, { name: 'bar' }).setOptions({ runValidators: true });
}
```